### PR TITLE
18 token movement monitoring

### PR DESCRIPTION
**Pull request by rpoletaev**
add func TokenTransferWithHistory
add model history
add monitoring funcs
add limit check to block_generator

### CODE REVIEW

File: packages/model/history.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	var res result
+	err = db.Table("1_history").Select("SUM(amount) as amount").
+		Where("created_at > NOW() - interval '24 hours' AND amount > 0").Scan(&res).Error
+
+	return res.Amount, err
+}
+
+// GetExcessFromToTokenMovementPerDay returns from to pairs where sum of amount greather than fromToPerDayLimit per 24 hours
+func GetExcessFromToTokenMovementPerDay() (excess []APLTransfer, err error) {
+	db := GetDB(nil)
```
Comment (potehinre) [inline packages/model/history.go]: better pass transaction as first parameter for that function, and then GetDB(transaction)

File: packages/model/history.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		Group("sender_id, recipient_id").
+		Having("SUM(amount) > ?", fromToPerDayLimit).
+		Scan(&excess).Error
+
+	return excess, err
+}
+
+// GetExcessTokenMovementQtyPerBlock returns from to pairs where APL transactions count greather than tokenMovementQtyPerBlockLimit per 24 hours
+func GetExcessTokenMovementQtyPerBlock(blockID int64) (excess []APLTransfer, err error) {
+	db := GetDB(nil)
```
Comment (potehinre) [inline packages/model/history.go]: Same stuff

File: packages/model/history.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+// APLTransfer from to amount
+type APLTransfer struct {
+	SenderID    int64
+	RecipientID int64
+	Amount      float64
+}
+
+// GetExcessCommonTokenMovementPerDay returns sum of amounts 24 hours
+func GetExcessCommonTokenMovementPerDay() (amount float64, err error) {
+	db := GetDB(nil)
```
Comment (potehinre) [inline packages/model/history.go]: same stuff

File: packages/notificator/token_movements.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	}
+	return err
+}
+
+// CheckTokenMovementLimits check all limits
+func CheckTokenMovementLimits(conf conf.TokenMovementConfig, blockID int64) {
+
+	amount, err := model.GetExcessCommonTokenMovementPerDay()
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Error("check common token movement")
```
Comment (potehinre) [inline packages/notificator/token_movements.go]: type: consts.DBError

File: packages/notificator/token_movements.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	} else if amount > networkPerDayLimit {
+		msg := fmt.Sprintf(networkPerDayMsgTemplate, amount)
+		sendEmail(conf, msg)
+	}
+
+	transfers, err := model.GetExcessFromToTokenMovementPerDay()
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Error("check from to token movement")
```
Comment (potehinre) [inline packages/notificator/token_movements.go]: same stuff

File: packages/notificator/token_movements.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	} else {
+		for _, transfer := range transfers {
+			msg := fmt.Sprintf(fromToDayLimitMsgTemplate, transfer.SenderID, transfer.RecipientID, transfer.Amount)
+			sendEmail(conf, msg)
+		}
+	}
+
+	transfers, err = model.GetExcessTokenMovementQtyPerBlock(blockID)
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Error("check token movement per block")
```
Comment (potehinre) [inline packages/notificator/token_movements.go]: Same stuff

**Review by potehinre — APPROVED** [review]

File: packages/model/history.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/shopspring/decimal"
+)
+
+const (
+	fromToPerDayLimit             = 10000
+	tokenMovementQtyPerBlockLimit = 100
```
Comment (ghost) [inline packages/model/history.go]: common consts should be moved to consts package

File: packages/notificator/token_movements.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+	"net/smtp"
+
+	"github.com/GenesisKernel/go-genesis/packages/conf"
+	"github.com/GenesisKernel/go-genesis/packages/model"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	networkPerDayLimit            = 100000000
```
Comment (ghost) [inline packages/notificator/token_movements.go:19]: common consts should be moved to consts package

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by potehinre — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**